### ISSUE TITLE
Kinematic body's transform origin is set to nan by move_and_slide

### ISSUE THREAD

**Issue by Norodix** [godotengine/godot#65179]
### Godot version

v3.5.stable.official [991bb6ac7]

### System information

Win10, Linux (popOS 22.04), AMD, Nvidia, Intel, Ryzen

### Issue description

In our 3d game we discovered an elusive bug where kinematic bodies' interaction causes the transforms to be set to nan. The error happens right after calling move_and_slide and both the returned velocity and the kinematic body's transforms' origin have all nan members.
This makes further calculations impossible using move_and_slide().

The error occurs rarely and I could not find anything strange in the input to move_and_slide or in the transform of the objects before the call. 

I noticed it triggers mostly when kinematic bodies are dropped onto each other, so the sample project repeatedly creates this situation.

We tested it on a windows machine with nvidia gpu and amd processor and on a linux machine with intel cpu and amd gpu and the issue seems to be independent of the platform. 

### Steps to reproduce

I managed to create a minimal project that exhibits the error. The code is a bit convoluted as it is a cut down part of the more complex movement code from our game, but it is less than 100 lines. 

To reproduce the bug simply run the project and wait. You might have to wait a few minutes or restart the project to trigger the bug, but this was the most consistent I could make it happen.

When the bug occurs the code should halt at the breakpoint.

### Minimal reproduction project

[KinematicBodyBugHunt.zip](https://github.com/godotengine/godot/files/9467345/KinematicBodyBugHunt.zip)

**Comment by Norodix** [comment]
Also, to trigger the error it seems to be necessary that the kinematic bodies move and rotate. That is why there is some random rotation included in the example project.

**Comment by Norodix** [comment]
Its hard to say definitively because it sometimes takes a long time to trigger the bug but I dont seem to be able to trigger it in 3.4.5

**Comment by Zireael07** [comment]
I saw the same error (nan values) in my game in 4.0 alpha 14. I fixed it by making very very sure cars don't spawn on top of each other... and never saw it again.

**Comment by Norodix** [comment]
It is hard to say definitively because its hard to consistently trigger the bug, but it seems to be related to a godot-physics related change from 3.4.5 to 3.5. It does not appear in 3.4.5 or when using 3.5 with bullet engine.

**Comment by Calinou** [comment]
@Norodix Can you reproduce this in any of the [3.5 betas and RCs](https://downloads.tuxfamily.org/godotengine/3.5/) to determine when the regression started?

**Comment by Norodix** [comment]
I managed to find which RC first included the regression! In RC6 I could not reproduce it in about half an hour of running, in RC7 I reproduced it in 1 minute. So it seems RC7 was the first where it occurred. 
I will look into finding the specific commit later today.

**Comment by Norodix** [comment]
I think I found it!

### Git bisect result
43e9816fb160cf0d1a6568f592963e03bea04f19 is the first bad commit
commit 43e9816fb160cf0d1a6568f592963e03bea04f19
Author: fabriceci <fabricecipolla@gmail.com>
Date:   Tue Jul 19 09:50:54 2022 +0200

    Fix the calculation of the angular velocity when the rotation speed is not high.
    
    (cherry picked from commit db7c91e0ada0c91ba48269cf3cc217631ca33bba)

 core/math/basis.cpp | 13 ++++++-------
 1 file changed, 6 insertions(+), 7 deletions(-)


### Git bisect logs

```
git bisect start
# good: [f05cecdc43ebfbe58ca9ed5b44fac888b4109dad] Merge pull request #55650 from kdiduk/pooled-arrays
git bisect good f05cecdc43ebfbe58ca9ed5b44fac888b4109dad
# bad: [38b95cc2fd7f898ead8c7be2732a6817eed1fba0] Merge pull request #63304 from hansemro/x11-eraser-case-insensitive-3.x
git bisect bad 38b95cc2fd7f898ead8c7be2732a6817eed1fba0
# good: [7e9f55c30d27d5ee5ce8eef8b7d7d0dd96b224ff] Merge pull request #63136 from akien-mga/3.x-gdnative-string-num-uint64
git bisect good 7e9f55c30d27d5ee5ce8eef8b7d7d0dd96b224ff
# good: [59629fdf20bdda235c87ad25750acc6fa5e68024] Merge pull request #49255 from kleonc/sprite_frames_editor-anim-search-box
git bisect good 59629fdf20bdda235c87ad25750acc6fa5e68024
# bad: [4ef99b409d32deade6f135627327dc984c212d91] Merge pull request #63163 from dsnopek/multiplayer-peer-custom-3.x
git bisect bad 4ef99b409d32deade6f135627327dc984c212d91
# bad: [558b96f1b183df2286b8d97c25efa2dbcd82c33e] Fix `CanvasItem` not exiting its canvas group on canvas exit
git bisect bad 558b96f1b183df2286b8d97c25efa2dbcd82c33e
# bad: [43e9816fb160cf0d1a6568f592963e03bea04f19] Fix the calculation of the angular velocity when the rotation speed is not high.
git bisect bad 43e9816fb160cf0d1a6568f592963e03bea04f19
# good: [4e1f30390d3eeb3a6db7cdeaf87635045fc47f1c] Use the right memory ordering in SafeNumeric operations
git bisect good 4e1f30390d3eeb3a6db7cdeaf87635045fc47f1c
# first bad commit: [43e9816fb160cf0d1a6568f592963e03bea04f19] Fix the calculation of the angular velocity when the rotation speed is not high.
```

**Comment by Norodix** [comment]
Seems to me the only difference is that `CMP_EPSILON` is defined to `0.00001` as opposed to the previously used `0.01`.

**Comment by Norodix** [comment]
@Calinou I verified it, by reverting the comparisons from `CMP_EPSILON` to `0.01` the issue was fixed.

Should I create a PR, or is it easier if a contributor fixes it? If so, which branch should I add it to so that it makes its way to both 4 and 3.x's next release?

**Comment by fabriceci** [comment]
@Norodix can you try if this PR solve your issue: https://github.com/godotengine/godot/pull/63428 ? (just need to replace the method in basis.cpp)

**Comment by Norodix** [comment]
@fabriceci I'm about to test it (its still building), but based on the code I'm not convinced its a full solution.

```
// As we have reached here there are no singularities so we can handle normally.
	double s = Math::sqrt((rows[2][1] - rows[1][2]) * (rows[2][1] - rows[1][2]) + (rows[0][2] - rows[2][0]) * (rows[0][2] - rows[2][0]) + (rows[1][0] - rows[0][1]) * (rows[1][0] - rows[0][1])); // Used to normalise.

	if (Math::abs(s) < CMP_EPSILON) {
		// Prevent divide by zero, should not happen if matrix is orthogonal and should be caught by singularity test above.
		s = 1;
	}

	x = (rows[2][1] - rows[1][2]) / s;
	y = (rows[0][2] - rows[2][0]) / s;
	z = (rows[1][0] - rows[0][1]) / s;

	r_axis = Vector3(x, y, z);
```
Based on my previous results it seems to me that when using `CMP_EPSILON` as the comparison for finding a singularity there are cases where a division by zero happens. Because of this I think there will be cases where we enter the "should not happen" section and s will be set to 1. This will make x, y, and z very small and the returned vector will not be a unit vector.

I dont think it is a correct behaviour. Maybe in the squareing and taking square roots causes the singularity to not be caught, but I have not 100% comprehended this part yet.

**Comment by Norodix** [comment]
I thought about it and tested it and it seems to be really about the acos function. Seems like the larger epsilon only hid the problem.

I could not use the PR because my test only runs on 3.x, but simply clamping the argument of acos solved the problem!

Can this fix make its way to 3.5?

**Comment by fabriceci** [comment]
@Norodix thanks for the testing! I think yes as this cause an issue, when the PR is merged into the master, there will be a cherry picking.
